### PR TITLE
Shbeniwa/native video impl

### PR DESCRIPTION
**Pull request by ShivangiBeniwal**
### Introduced Select Media V2 where partner apps can now record video in non-fullscreen mode

- This feature is natively supported by both ios and android platforms and doesn't go through the traditional lens flow.
- Following were the requirement for the feature : 
1) An attribute that indicates if the video needs to be launched in Picture In Picture mode or Full-Screen mode.
2) An Attribute to control the visibility of the STOP button in the recorder bar.
3) Channel for partner apps to listen for **Recording Started** event notification sent by host client.
4) Channel for partner apps to notify the host client to **Stop Recording**.

- Changes made in JS SDK MEDIA FILE :
1) Introduced an attribute in VideoProps called `isFullScreenMode`, default is TRUE
2) Introduced an attribute in VideoProps called `isStopButtonVisible`, default is TRUE
3) Implement the VideoControllerCallback and pass that in the object of VideoController in VideoProps
    `Select Media API call will now receive partial responses from the host client-side` and JS SDK layer will notify the partner app by invoking the respective VideoControllerCallback method if the app has registered.
4)  A new API called Media Controller has been introduced, to notify the host client about the event that partner apps want to notify. 
    `VideoController is a wrapper class on top of MediaController for easy and direct usage` of basic functionalities specific to video like stop(), resume(), pause() methods. 

> BI-DIRECTIONAL COMMUNICATION CHANNEL can be used by partner apps to listen to notifications about the ongoing recording flow and also communicate platform to control the experience programmatically at run-time !!

_We have designed keeping in mind the future scalability of this feature to support other media types like audio with minimal changes in JS SDK and no backward compatibility concerns._

**Note** - This is a mobile feature only and won't be implemented on desktop. 
Attached are the docs for current changes in JS SDK PR and a detailed design document in the channel conversation.

### CODE REVIEW

File: src/public/media.ts [hunk 1]
```diff
@@ -451,18 +560,32 @@ export namespace media {
...
-      }
-      const mediaArray: Media[] = [];
-      for (const attachment of localAttachments) {
-        mediaArray.push(new Media(attachment));
-      }
-      callback(err, mediaArray);
-    });
+    // What comes back from native as attachments would just be objects and will be missing getMedia method on them
+    sendMessageToParent(
+      'selectMedia',
```
Comment (ydogandjiev) [inline src/public/media.ts:565]: This name should always include the namespace/module so in this case it should be `media.selectMedia` and not just `selectMedia`

**Review by ydogandjiev — CHANGES_REQUESTED** [review]
Please setup time to review this feature and this PR with @sachingupta and myself. There are way too many changes in this file with a bunch of new patterns being established and I have way too many questions to do this in PR.

File: src/public/media.ts [hunk 2]
```diff
@@ -10,16 +10,19 @@ import {
...
 import {
   captureImageMobileSupportVersion,
   mediaAPISupportVersion,
   getMediaCallbackSupportVersion,
   scanBarCodeAPIMobileSupportVersion,
-  videoAndImageMediaAPISupportVersion,
+  nonFullScreenVideoModeAPISupportVersion,
 } from '../internal/constants';
 
 export namespace media {
```
Comment (sachingupta) [inline src/public/media.ts:28]: Please ask @anandvnath or @SonalJha to review this PR.

**Review by sachingupta — CHANGES_REQUESTED** [review]

**Review by praxman — APPROVED** [review]

File: src/public/media.ts (hunk 1 above)
Comment (ShivangiBeniwal) [reply src/public/media.ts:592]: As discussed on the call, will take all these changes in the next clean-up PR. 
Task 2096888: [JS SDK] Clean all the API signatures to have the namespace prefix

File: src/public/media.ts (hunk 2 above)
Comment (ShivangiBeniwal) [reply src/public/media.ts:26]: They are no longer part of the DeviceCapability team, @praxman, and @cpjoshi please review and approve if not done. Thanks.

File: src/internal/constants.ts [hunk 3]
```diff
@@ -13,6 +13,11 @@ export const defaultSDKVersionForCompatCheck = '2.0.1';
  */
 export const videoAndImageMediaAPISupportVersion = '2.0.2';
 
+/**
+ * This is the SDK version when selectMedia API - Video with non-full screen mode is supported on mobile.
```
[the remark this answers is not in the document]
Comment (ShivangiBeniwal) [reply src/internal/constants.ts]: Done.

File: src/internal/mediaUtil.ts [hunk 4]
```diff
@@ -1,5 +1,40 @@
...
+  }
+  return false;
+}
+
+/**
+ * Helper function which indicates if current API is supported on mobile or not.
+ * @returns SdkError if host client is not android/ios or if the requiredVersion is not
+ *          supported by platform or not. Null is returned in case of success.
+ */
+export function isApiSupportedOnMobile(requiredVersion: string = defaultSDKVersionForCompatCheck): SdkError {
```
[the remark this answers is not in the document]
Comment (ShivangiBeniwal) [reply src/internal/mediaUtil.ts]: Done.

File: src/private/interfaces.ts [hunk 5]
```diff
@@ -164,3 +166,35 @@ export interface UserJoinedTeamsInformation {
    */
   userJoinedTeams: TeamInformation[];
 }
+
```
[the remark this answers is not in the document]
Comment (ShivangiBeniwal) [reply src/private/interfaces.ts]: Cherry-picking the corresponding change in Android is not possible, hence will be taking up this comment in Dec sprint. 
Task 2096888: [JS SDK] Clean all the API signatures to have the namespace prefix

File: src/internal/mediaUtil.ts [hunk 6]
```diff
@@ -1,5 +1,40 @@
...
+  defaultSDKVersionForCompatCheck,
+  videoAndImageMediaAPISupportVersion,
+  nonFullScreenVideoModeAPISupportVersion,
+} from './constants';
+
+/**
+ * Helper function to identify if host client is either android or ios
+ */
+export function isHostClientMobile(): boolean {
+  if (GlobalVars.hostClientType == HostClientType.android || GlobalVars.hostClientType == HostClientType.ios) {
```
[the remark this answers is not in the document]
Comment (ShivangiBeniwal) [reply src/internal/mediaUtil.ts]: Done.

File: src/internal/mediaUtil.ts [hunk 7]
```diff
@@ -46,6 +81,32 @@ export function decodeAttachment(attachment: media.MediaChunk, mimeType: string)
...
+    return isApiSupportedOnMobile(nonFullScreenVideoModeAPISupportVersion);
+  }
+  return null;
+}
+
+/**
+ * Function returns true if the app has registered to listen to video controller events, else false.
+ */
+export function isVideoControllerRegistered(mediaInputs: media.MediaInputs): boolean {
+  if (
```
[the remark this answers is not in the document]
Comment (ShivangiBeniwal) [reply src/internal/mediaUtil.ts:68]: This is not possible as we are checking if videoProps && videoController are not null or undefined and the compiler gets confused to decide the return type as the condition gets converted to `boolean | videoprops | videocontroller`

File: src/internal/mediaUtil.ts [hunk 8]
```diff
@@ -68,6 +129,22 @@ export function isMediaCallForVideoAndImageInputs(mediaInputs: media.MediaInputs
   return false;
 }
 
+/**
+ * Returns true if the mediaInput params are called for non-full screen video mode and false otherwise
+ */
+export function isMediaCallForNonFullScreenVideoMode(mediaInputs: media.MediaInputs): boolean {
+  if (mediaInputs) {
+    if (
```
[the remark this answers is not in the document]
Comment (ShivangiBeniwal) [reply src/internal/mediaUtil.ts]: Same compiler issue as above.

File: src/public/media.ts [hunk 9]
```diff
@@ -316,12 +340,103 @@ export namespace media {
...
+      const err = isApiSupportedOnMobile(nonFullScreenVideoModeAPISupportVersion);
+      if (err) {
+        if (callback) {
+          callback(err);
+        }
+        return;
+      }
+
+      const params: MediaControllerParam = { mediaType: this.getMediaType(), mediaControllerEvent: mediaEvent };
+      sendMessageToParent('mediaController', [params], (err?: SdkError) => {
```
[the remark this answers is not in the document]
Comment (ShivangiBeniwal) [reply src/public/media.ts:391]: Cherry-picking the corresponding change in Android is not possible, hence will be taking up this comment in Dec sprint. 
Task 2096888: [JS SDK] Clean all the API signatures to have the namespace prefix

**Review by sachingupta — APPROVED** [review]

**Review by ydogandjiev — APPROVED** [review]

**Pull request merged**